process/peer/process.go

```diff
@@ -345,6 +389,123 @@ func (p *validatorStatistics) getPeerAccount(address []byte) (state.PeerAccountH
 	return peerAccount, nil
 }
 
+// loadPreviousShardHeaders loads the previous shard headers for a given metablock. For the metachain it's easy
+//  since it has all the shard headers in its storage, but for the shard it's a bit trickier and we need
+//  to iterate through past metachain headers until we find all the ShardData we are interested in
+func (p *validatorStatistics) loadPreviousShardHeaders(currentHeader, previousHeader *block.MetaBlock) error {
+
+	p.mutPrevShardInfo.Lock()
+	defer p.mutPrevShardInfo.Unlock()
+
+	p.prevShardInfo = make(map[uint32]block.ShardData)
+	missingPreviousShardData := make(map[uint32]block.ShardData)
+
+	for _, currentShardData := range currentHeader.ShardInfo {
+
+		if currentShardData.Nonce == 1 {
+			continue
+		}
+
+		prevShardData := p.getMatchingPrevShardData(currentShardData, currentHeader.ShardInfo)
+		if prevShardData != nil {
+			p.prevShardInfo[currentShardData.ShardID] = *prevShardData
+			continue
+		}
+
+		prevShardData = p.getMatchingPrevShardData(currentShardData, previousHeader.ShardInfo)
+		if prevShardData != nil {
+			p.prevShardInfo[currentShardData.ShardID] = *prevShardData
+			continue
+		}
+
+		missingPreviousShardData[currentShardData.ShardID] = currentShardData
+	}
+
+	searchHeader := &block.MetaBlock{}
```

Comment:
you can extract also this block from line 424 to 445 into a function

Reply:
done